---
title: "Plotting expression levels"
author: "Jared J. Homola"
date: "4/23/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message = FALSE, warning = FALSE, echo = FALSE, results = 'hide'}
## Load packages and data
library(ggrepel)
library(edgeR)
library(RColorBrewer)
library(gplots)
library(kableExtra)
library(gridExtra)
library(tidyverse)

setwd("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/")

## Load data
annot <- read.delim("./extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt <- read.table("./extData/rsem.isoform.counts.matrix", header = TRUE)
covars <- read.csv("./extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)

## Clean data
cnt.bursa <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Bursa"))

cnt.ileum <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Ileum"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript_id, pt1, pt2, gene, isoform)

#Convert to DGEList object
dge.bursa <- DGEList(counts=cnt.bursa)
dge.bursa$genes <- annot[match(rownames(dge.bursa$counts), annot$transcript_id),]
#CPM and log-CPM
cpm.bursa <- cpm(dge.bursa)
lcpm.bursa <- cpm(dge.bursa, log = TRUE)

#Convert to DGEList object
dge.ileum <- DGEList(counts=cnt.ileum)
dge.ileum$genes <- annot[match(rownames(dge.ileum$counts), annot$transcript_id),]
#CPM and log-CPM
cpm.ileum <- cpm(dge.ileum)
lcpm.ileum <- cpm(dge.ileum, log = TRUE)

## Master lcpm tib
lcpm.bursa.tmp <- lcpm.bursa %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript")
lcpm.ileum.tmp <- lcpm.ileum %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript")

lcpm <- lcpm.bursa.tmp %>% 
  full_join(lcpm.ileum.tmp) %>% 
  replace(., is.na(.), "0")
```

### Plotting single transcripts
```{r, message = FALSE, warning = FALSE, echo = FALSE}
targetTrans <- "DN6338_c0_g1_i3"
lcpm %>%
  filter(transcript %in% targetTrans) %>% 
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "expression") %>% 
  separate(sample, into = c(NA, NA, "Tissue", "bird", NA)) %>% 
  mutate(bird = as.integer(bird)) %>% 
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "CONTROL", "LOW", "MODERATE", "HIGH")) %>% 
  ggplot(aes(y = expression, x = factor(SSgroup.virus.sac), fill = factor(Tissue))) +
  ylab("Log(Counts per million)") +
  xlab("Virus shedding group on day 1") +
  ggtitle(paste("Transcript: ", targetTrans)) +
  geom_boxplot() +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank())
```
  
### Single transcripts 
The line is drawn as a Loess fuction and gray area shows 95% confidence intervals.
```{r, message = FALSE, warning = FALSE, echo = FALSE}
targetTrans <- "DN6338_c0_g1_i3"
lcpm %>%
  filter(transcript %in% targetTrans) %>% 
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "lcpm") %>% 
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>% 
  mutate(bird = as.integer(bird)) %>% 
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  ggplot(aes(x = log(virus.sac+0.01), y = lcpm, color = tissue)) +
  geom_point() +
  geom_smooth(method = "loess", span = 1.0) +
  ylab("Log(Counts per million)") +
  xlab("Log(Viral titer on day of sacrifice)") +
  ggtitle(paste("Transcript: ", targetTrans)) +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank())
```
  

### Plotting multiple transcripts/genes
#### Categorical x axis, line plot

This example is for all genes with a SwissProt annotation for MX protein encoding genes. Note that these plots put all associated transcripts on one figure. 
```{r, message = FALSE, warning = FALSE, echo = FALSE}
ctlInf.figure <- function(queryGene, ...) {
  
  ## Get list of transcripts associated with a gene name
  targetTrans <- annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, NA, NA), "\\^") %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  filter(SwissProt_GeneName == queryGene) %>% 
  distinct(transcript_id, .keep_all = TRUE)

  ## Create a vector of those transcripts
  targetTrans <- targetTrans$transcript_id
  
  ## Create the tibble that will be plotted 
  plot.dat <- lcpm %>%
  filter(transcript %in% targetTrans) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "expression") %>%
  separate(sample, into = c(NA, NA, "Tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "CONTROL", "LOW", "MODERATE", "HIGH")) %>%
  arrange(SSgroup.virus.sac) %>%
  mutate(
    xVar = paste(SSgroup.virus.sac, "-", bird),
    xVar = factor(xVar, levels = unique(xVar))
  ) %>%
  select(transcript, Tissue, bird, expression, xVar)
  
  ## Subset plotting tibble into bursa and ileum tibbles
  plot.dat.ileum <- plot.dat %>% filter(Tissue == "Ileum")
  plot.dat.bursa <- plot.dat %>% filter(Tissue == "Bursa")

  ## Generate plot
  plot <- ggplot(data = plot.dat.ileum, aes(y = expression, x = xVar, group = transcript)) +
  geom_line(color = "red", alpha = 0.1) +
  geom_line(data = plot.dat.bursa,
    aes(y = expression, x = xVar, group = transcript),
    color = "blue", alpha = 0.1) +
  stat_summary(aes(y = expression, group = 1),
    fun.y = mean, colour = "red", geom = "line", group = 1, lwd = 2) +
  stat_summary(data = plot.dat.bursa,
    aes(y = expression, group = 1),
    fun.y = mean, colour = "blue", geom = "line", group = 1, lwd = 2) +
  ggtitle(paste("Transcripts annotated as", queryGene)) +
  geom_rect(aes(xmin = 0.4, xmax = 2.6, ymin = 8.5, ymax = 10.5),
  color = "black", fill = "white") +
  annotate("Text", x = 1.5, y = 10, label = "Bursa", color = "blue", fontface = 2,size = 5) +
  annotate("Text", x = 1.5, y = 9.0, label = "Ileum",
    color = "red", fontface = 2, size = 5) +
  ylab("Log(Counts per million)") +
  xlab("Bird") +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
  
  print(plot)
}

ctlInf.figure("MX")

```

### Plotting multiple transcripts/genes with a given annotation
#### Continuous x axis, separate scatterplots for each transcript
Same data as above, just plotted on a continuous axis and in separate plots for each transcript. Trend lines are plotted via Loess smoothing.

```{r, message = FALSE, warning = FALSE, fig.height = 25, fig.width = 10, echo = FALSE}
targetTrans <- annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, NA, NA), "\\^") %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  filter(SwissProt_GeneName == "MX") %>% 
  distinct(transcript_id, .keep_all = TRUE)

covars.sub <- covars %>% 
  filter(group == "I1" | group == "I3") %>% 
  select(bird, virus.sac) %>% 
  mutate(bird = sprintf("%02d", bird)) %>% 
  mutate(subSam = paste0("_", bird, "_"))
  
tmp.lcpm <- lcpm %>% 
  select(contains(covars.sub$subSam)) %>%
  mutate(transcript_id = lcpm$transcript) %>% 
  filter(transcript_id %in% targetTrans$transcript_id) %>% 
  pivot_longer(cols = starts_with("align"),
               names_to = "sample",
               values_to = "lcpm") %>% 
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>% 
  left_join(covars.sub)


for(gene in unique(tmp.lcpm$transcript_id)) {
  name <- paste0("plot.", gene)
  assign(name, tmp.lcpm %>%
      filter(transcript_id == gene) %>% 
      ggplot(aes(x = log(virus.sac), y = lcpm, color = tissue)) +
      geom_point() +
      geom_smooth(method = "loess", span = 1.5) +
      ylab("Log(Counts per million)") +
      xlab("Log(Viral titer on day of sacrifice)") +
      ggtitle(gene) +
      theme_classic() +
      theme(legend.title = element_blank(),
            axis.text.x = element_text(angle = 45, hjust = 1)))
}

grid.arrange(plot.DN110170_c0_g1_i1, plot.DN2085_c0_g2_i1,
             plot.DN2085_c0_g2_i2, plot.DN2085_c0_g2_i5,
             plot.DN2085_c0_g2_i6, plot.DN2085_c0_g2_i7,
             plot.DN2085_c0_g2_i10, plot.DN2085_c0_g2_i11,
             plot.DN2085_c0_g2_i12, plot.DN2085_c0_g2_i13,
             plot.DN2085_c0_g2_i14, plot.DN2085_c0_g2_i15,
             plot.DN2085_c0_g2_i16, nrow = 5)
```

