---
title: "Candidate Gene Analysis, Correlation method"
author: "Jared J. Homola"
date: "7/14/2020"
output: html_document
---

The results illustrated here are from running Pearson correlations on all candidate genes and transcripts to evaluate relationships between viral titer (both sacrifice day and mean titer) and gene/transcript expression. Both variables were log transformed prior to analysis because a high titer outlier was driving most conclusions otherwise. 

A false discovery rate q value of 0.05 was used to determine significance and I further filtered to include only correlation coefficients >0.6 or <-0.6. There were initially several results with significant p-values that were driven by only a couple samples. Filtering on both correlation coefficients and p-values removes these potentially spurious correlations.

The table below shows transcripts/genes that met filtering criteria. There were no retained results for mean viral titers. The comparison column indicates which tissue type (B/Bursa or I/Ileum) and which analysis level was significant (G/Gene or T/Transcript).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/")
load("aPrioriGenes-Corr-log.Rws")

```

```{r message = FALSE, warning = FALSE, echo = FALSE, results = 'hide'}
## Load packages and data
library(tidyverse)
library(kableExtra)
library(edgeR)
library(ggrepel)

## Load data
annot <- read.delim("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt.trans <- read.table("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/rsem.isoform.counts.matrix", header = TRUE)
cnt.gene <- read.table("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/rsem.gene.counts.matrix", header = TRUE)
covars <- read.csv("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)
targets <- read_delim("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/aPrioriTranscripts_V2.csv", delim = ",") %>%
  filter(include == "Yes" | include == "yes")

## Clean data
cnt.bursa.gene <- cnt.gene %>%
  select(-alignEstimateAbundance_BWTE_Ileum_36_S50,
         -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
         -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("gene") %>%
  separate(gene, into = c(NA, "pt1", "pt2", "gene", NA)) %>%
  unite(gene, pt1, pt2, gene) %>%
  column_to_rownames("gene") %>%
  select(contains("Bursa"))

cnt.ileum.gene <- cnt.gene %>%
  select(-alignEstimateAbundance_BWTE_Ileum_36_S50,
         -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
         -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("gene") %>%
  separate(gene, into = c(NA, "pt1", "pt2", "gene", NA)) %>%
  unite(gene, pt1, pt2, gene) %>%
  column_to_rownames("gene") %>%
  select(contains("Ileum"))

cnt.bursa.trans <- cnt.trans %>%
  select(-alignEstimateAbundance_BWTE_Ileum_36_S50,
         -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
         -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>%
  select(contains("Bursa"))

cnt.ileum.trans <- cnt.trans %>%
  select(-alignEstimateAbundance_BWTE_Ileum_36_S50,
         -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
         -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>%
  select(contains("Ileum"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(gene_id, pt1, pt2, gene, remove = FALSE) %>%
  unite(transcript_id, pt1, pt2, gene, isoform) %>% 
  filter(transcript_id %in% targets$transcript_id | gene_id %in% targets$gene_id)

#### Calculate log(CPM) and assemble master DFs ####
#Convert to DGEList object
dge.bursa.trans <- DGEList(counts=cnt.bursa.trans)
dge.bursa.gene <- DGEList(counts=cnt.bursa.gene)
dge.ileum.trans <- DGEList(counts=cnt.ileum.trans)
dge.ileum.gene <- DGEList(counts=cnt.ileum.gene)

#CPM and log-CPM
lcpm.bursa.trans <- cpm(dge.bursa.trans, log = TRUE)
lcpm.bursa.gene <- cpm(dge.bursa.gene, log = TRUE)
lcpm.ileum.trans <- cpm(dge.ileum.trans, log = TRUE)
lcpm.ileum.gene <- cpm(dge.ileum.gene, log = TRUE)

## Master lcpm tibs
# Trans
lcpm.bursa.tmp <- lcpm.bursa.trans %>%
  as_tibble(rownames = NA) %>%
  rownames_to_column("transcript")
lcpm.ileum.tmp <- lcpm.ileum.trans %>%
  as_tibble(rownames = NA) %>%
  rownames_to_column("transcript")

lcpm.trans <- lcpm.bursa.tmp %>%
  full_join(lcpm.ileum.tmp) %>%
  replace(., is.na(.), "0") %>%
  filter(transcript %in% targets$transcript_id) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "lcpm") %>%
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  mutate(levelGT = "transcript", identifier = transcript) %>%
  select(identifier, levelGT, tissue, bird, lcpm, virus.sac, group, virus.dpi1, virus.dpi2, virus.dpi3, virus.dpi4, virus.dpi5) %>%
  mutate(log.virus.sac = log10(virus.sac+1)) %>%
  pivot_longer(cols = virus.dpi1:virus.dpi5,
               names_to = "dpi",
               values_to = "titer") %>%
  group_by(identifier, bird, levelGT, tissue, lcpm, log.virus.sac, group) %>%
  summarize(log.meanTiter1to5 = log10(mean(titer, na.rm = TRUE)))

# Gene
lcpm.bursa.tmp <- lcpm.bursa.gene %>%
  as_tibble(rownames = NA) %>%
  rownames_to_column("gene")
lcpm.ileum.tmp <- lcpm.ileum.gene %>%
  as_tibble(rownames = NA) %>%
  rownames_to_column("gene")

lcpm.all <- lcpm.bursa.tmp %>%
  full_join(lcpm.ileum.tmp) %>%
  replace(., is.na(.), "0")  %>%
  filter(gene %in% targets$gene_id) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "lcpm") %>%
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  mutate(levelGT = "gene", identifier = gene) %>%
  select(identifier, levelGT, tissue, bird, lcpm, virus.sac, group, virus.dpi1, virus.dpi2, virus.dpi3, virus.dpi4, virus.dpi5) %>%
  mutate(log.virus.sac = log10(virus.sac+1)) %>%
  pivot_longer(cols = virus.dpi1:virus.dpi5,
               names_to = "dpi",
               values_to = "titer") %>%
  group_by(identifier, bird, levelGT, tissue, lcpm, log.virus.sac, group) %>%
  summarize(log.meanTiter1to5 = log10(mean(titer + 1, na.rm = TRUE))) %>%
  bind_rows(lcpm.trans) %>% 
  mutate(group = recode(group,
                        C1 = "Ctl",
                        C14 = "Ctl")) %>% 
  mutate(group = factor(group, 
                          levels = c("Ctl", "I1", "I3", "I5", "I14")))

#### Gene query database ####
annot.all <- annot %>%
  select(transcript_id,
         gene_id,
         sprot_Top_BLASTX_hit,
         sprot_Top_BLASTP_hit,
         gene_ontology_BLASTX,
         gene_ontology_BLASTP,
         Kegg,
         eggnog,
         Pfam) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot_geneName_BlastX", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%
  separate(sprot2, sep = "=", into = c(NA, "sprot_geneFunction_BlastX")) %>%
  separate(sprot_geneFunction_BlastX, sep = ";", into = c("sprot_geneFunction_BlastX", NA)) %>%
  separate(sprot_Top_BLASTP_hit, into = c("sprot_geneName_BlastP", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%
  separate(sprot2, sep = "=", into = c(NA, "sprot_geneFunction_BlastP")) %>%
  separate(sprot_geneFunction_BlastP, sep = ";", into = c("sprot_geneFunction_BlastP", NA)) %>%
  separate(gene_ontology_BLASTX, sep = "\\`", into = paste("GO_BlastX", 1:5, sep = "_"), extra = "drop", fill = "right") %>%
  separate(gene_ontology_BLASTP, sep = "\\`", into = paste("GO_BlastP", 1:5, sep = "_"), extra = "drop", fill = "right") %>%
  separate(Pfam, sep = "\\`", into = paste("Pfam", 1:5, sep = "_"), extra = "drop", fill = "right") %>%
  select(transcript_id, gene_id, sprot_geneName_BlastX, sprot_geneFunction_BlastX) %>% 
  as_tibble()
```

## Summarize results
```{r, message = FALSE, warning = FALSE, echo = FALSE}
results.table <- sigResults.mean %>% 
  bind_rows(sigResults.sac) %>% 
  filter(Est < -0.6 | Est > 0.6)

kable(results.table) %>%
  kable_styling("striped", full_width = F)

```


## Functions for analyzing and plotting data
```{r, message = FALSE, warning = FALSE, echo = FALSE}
correlationPlot.mean <- function(target, targetTissue, targetLevel, ...) {
  annotation <- annot.all %>%
    filter(transcript_id == target | gene_id == target) %>%
    filter(sprot_geneName_BlastX != ".") %>%
    select(sprot_geneName_BlastX) %>%
    slice(1)
  
  plot <- lcpm.all %>%
    filter(levelGT == targetLevel,
           tissue == targetTissue) %>% 
    filter(identifier == target) %>% 
    ggplot(aes(x = log.meanTiter1to5, y = lcpm)) +
    geom_point(aes(shape = group, color = group), size = 5) +
    ylab("Log2(Counts per million)") +
    xlab("Log10(Average viral titer + 1)") +
    scale_shape_manual(values=c(13, 15:18)) +
    guides(shape = guide_legend(override.aes = list(size = 5))) +
    theme_classic() +
    labs(title=paste0(annotation[1], " - ", target),
         subtitle=paste0(targetTissue, " - ", targetLevel,
                         "; Adj. p = ", round(targetInfo$adj.p.value, 5),
                         "; Pearson correlation coefficent = ", round(targetInfo$Est, 3)))
  print(plot)
}


correlationPlot.sac <- function(target, targetTissue, targetLevel, ...) {
  annotation <- annot.all %>%
    filter(transcript_id == target | gene_id == target) %>%
    filter(sprot_geneName_BlastX != ".") %>%
    select(sprot_geneName_BlastX) %>%
    slice(1)
  
  plot <- lcpm.all %>%
    filter(levelGT == targetLevel,
           tissue == targetTissue) %>% 
    filter(identifier == target) %>% 
    ggplot(aes(x = log.virus.sac, y = lcpm)) +
    geom_point(aes(shape = group, color = group), size = 5) +
    ylab("Log2(Counts per million)") +
    xlab("Log10(Sacrifice day viral titer + 1)") +
    scale_shape_manual(values=c(13, 15:18)) +
    guides(shape = guide_legend(override.aes = list(size = 5))) +
    theme_classic() +
    labs(title=paste0(annotation[1], " - ", target),
         subtitle=paste0(targetTissue, " - ", targetLevel,
                         "; Adj. p = ", round(targetInfo$adj.p.value, 5),
                         "; Pearson correlation coefficent = ", round(targetInfo$Est, 3)))
  print(plot)
}

```

## Plotting significant results for sacrifice day titer
```{r, message = FALSE, warning = FALSE, echo = FALSE}
for(z in 1:nrow(results.table)) {
  targetInfo <- results.table %>% 
    slice(z) %>% 
    separate(comparison, c("metric", "tissueLevel")) %>% 
    mutate(tissue = if_else(tissueLevel == "BG" | tissueLevel == "BT", "Bursa", "Ileum")) %>% 
    mutate(level = if_else(tissueLevel == "BG" | tissueLevel == "IG", "gene", "transcript"))

  if(targetInfo$metric == "sac") {
    correlationPlot.sac(targetInfo$identifier, targetInfo$tissue, targetInfo$level)
  }
  else {
    correlationPlot.mean(targetInfo$identifier, targetInfo$tissue, targetInfo$level)
  }
}
```
