---
title: "Plotting expression levels"
author: "Jared J. Homola"
date: "4/23/2020"
output: html_document
---
This document provides ideas for plotting gene/transcript expression levels. A few notes...  
  
1) Amanda and I are thinking of the manuscript's analyses in two broad categories: a) Differential expression analyses and b) Analyses of a priori listed genes of interest. This document is organized by those categories.
2) Each "scenario" given below is an example of an instance where we will likely want to plot data.  
3) The example given for a differential expression analysis is for the day 1 control, low, moderate, and high shedding birds at the transcript level.  
4) When plotting expression levels for each of these, we could stick to our low-moderate-high shedding level categories or use a continuous axis of virus titer. If we use a continuous axis, we could highlight breakpoints along it where we propose low, moderate, and high levels exist so it could still be tied back to the DE analyses, but that is not shown here.  
5) For DE plots, we can attach a p-value from the DE analysis to these plots. For analysis of a priori genes, we do not do a statistically test at this point, instead just looking at the overall pattern.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message = FALSE, warning = FALSE, echo = FALSE, results = 'hide'}
## Load packages and data
library(ggrepel)
library(edgeR)
library(RColorBrewer)
library(gplots)
library(kableExtra)
library(gridExtra)
library(tidyverse)

setwd("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/")

## Load data
annot <- read.delim("./extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt <- read.table("./extData/rsem.isoform.counts.matrix", header = TRUE)
covars <- read.csv("./extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)

## Clean data
cnt.bursa <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Bursa"))

cnt.ileum <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Ileum"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript_id, pt1, pt2, gene, isoform)

### Set up model effects
bird <- as.factor(covars$bird)
sex <- as.factor(covars$sex)
age <- as.numeric(covars$age)
weight <- covars$wt_55
group <- as.factor(covars$group)
pool.bursa <- as.factor(covars$Pool.Bursa)
pool.ileum <- as.factor(covars$Pool.Ileum)
ssgroup.virus.sac <- as.factor(covars$SSgroup.virus.sac)

covars.tib <- data.frame(bird,
                    sex,
                    age,
                    weight,
                    group,
                    pool.bursa,
                    pool.ileum,
                    ssgroup.virus.sac) %>% 
  as_tibble()
#Convert to DGEList object
dge.bursa <- DGEList(counts=cnt.bursa)
dge.bursa$genes <- annot[match(rownames(dge.bursa$counts), annot$transcript_id),]

#CPM and log-CPM
cpm.bursa <- cpm(dge.bursa)
lcpm.bursa <- cpm(dge.bursa, log = TRUE)

#### Retain only transcripts expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge.bursa$counts==0) == length(cnt.bursa[1,]))
keep.exprs <- rowSums(cpm.bursa>0.5) >= length(cnt.bursa[1,])/4

dge.bursa <- dge.bursa[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge.bursa$counts==0) == length(cnt.bursa[1,]))

#TMM normalization
dge.bursa <- calcNormFactors(dge.bursa, method = "TMM")

#Convert to DGEList object
dge.ileum <- DGEList(counts=cnt.ileum)
dge.ileum$genes <- annot[match(rownames(dge.ileum$counts), annot$transcript_id),]

#CPM and log-CPM
cpm.ileum <- cpm(dge.ileum)
lcpm.ileum <- cpm(dge.ileum, log = TRUE)

#### Retain only transcript expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge.ileum$counts==0) == length(cnt.ileum[1,]))
keep.exprs <- rowSums(cpm.ileum>0.5) >= length(cnt.ileum[1,])/4

dge.ileum <- dge.ileum[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge.ileum$counts==0) == length(cnt.ileum[1,]))

#TMM normalization
dge.ileum <- calcNormFactors(dge.ileum, method = "TMM")

#Assign groups in DGElist
dge.bursa$samples$group <- covars.tib$group
dge.ileum$samples$group <- covars.tib$group

### Divide out day 1
dge.bursa.1 <- dge.bursa[,grep('\\bC1\\b|\\bI1\\b', dge.bursa$samples$group)]
dge.ileum.1 <- dge.ileum[,grep('\\bC1\\b|\\bI1\\b', dge.ileum$samples$group)]
covars.1 <- covars.tib %>% filter(group %in% c("C1", "I1")) %>% droplevels()

## Master lcpm tib
lcpm.bursa.tmp <-
  lcpm.bursa %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript")
lcpm.ileum.tmp <-
  lcpm.ileum %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript")

lcpm <-
  lcpm.bursa.tmp %>% 
  full_join(lcpm.ileum.tmp) %>% 
  replace(., is.na(.), "0")


## Bursa
design.bursa = model.matrix(~ 0 +
                        covars.1$ssgroup.virus.sac +
                        covars.1$age +
                        covars.1$sex + 
                        covars.1$weight + 
                        covars.1$pool.bursa)

colnames(design.bursa) <- c("CONTROL", "HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool2", "pool3")

contr.matrix.bursa = makeContrasts(
  CvL = CONTROL - LOW,
  CvM = CONTROL - MODERATE,
  CvH = CONTROL - HIGH,
  LvM = LOW - MODERATE,
  LvH = LOW - HIGH,
  MvH = MODERATE - HIGH,
  levels = colnames(design.bursa))

v.bursa <- voomWithQualityWeights(dge.bursa.1, design.bursa)
vfit.bursa <- lmFit(v.bursa, design.bursa) 
vfit.bursa <- contrasts.fit(vfit.bursa, contrasts = contr.matrix.bursa)
tfit.bursa <- treat(vfit.bursa, lfc = 1.0)

## Ileum
design.ileum = model.matrix(~ 0 +
                        covars.1$ssgroup.virus.sac +
                        covars.1$age +
                        covars.1$sex + 
                        covars.1$weight + 
                        covars.1$pool.ileum)

colnames(design.ileum) <- c("CONTROL", "HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool3")

contr.matrix.ileum = makeContrasts(
  CvL = CONTROL - LOW,
  CvM = CONTROL - MODERATE,
  CvH = CONTROL - HIGH,
  LvM = LOW - MODERATE,
  LvH = LOW - HIGH,
  MvH = MODERATE - HIGH,
  levels = colnames(design.ileum))

v.ileum <- voomWithQualityWeights(dge.ileum.1, design.ileum)
vfit.ileum <- lmFit(v.ileum, design.ileum) 
vfit.ileum <- contrasts.fit(vfit.ileum, contrasts = contr.matrix.ileum)
tfit.ileum <- treat(vfit.ileum, lfc = 1.0)

## Summarize results
dt.bursa <- decideTests(tfit.bursa, p.value = 0.1, adjust.method = "fdr")

dt.tib.bursa <- as_tibble(dt.bursa, rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  mutate_at(vars(contains("v")), as.numeric) %>% 
  rename(CvL.bursa = CvL,
         CvM.bursa = CvM,
         CvH.bursa = CvH,
         LvM.bursa = LvM,
         LvH.bursa = LvH,
         MvH.bursa = MvH)  %>% 
  filter_at(vars(CvL.bursa:MvH.bursa), any_vars(. != 0))

dt.ileum <- decideTests(tfit.ileum, p.value = 0.1, adjust.method = "fdr")

dt.tib.ileum <- as_tibble(dt.ileum, rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  mutate_at(vars(contains("v")), as.numeric) %>% 
  rename(CvL.ileum = CvL,
         CvM.ileum = CvM,
         CvH.ileum = CvH,
         LvM.ileum = LvM,
         LvH.ileum = LvH,
         MvH.ileum = MvH) %>% 
  filter_at(vars(CvL.ileum:MvH.ileum), any_vars(. != 0))

dt.tib <- dt.tib.bursa %>% 
  full_join(dt.tib.ileum) %>% 
  replace(., is.na(.), "0") %>% 
  mutate_at(vars(contains("v")), as.numeric)


```

```{r, message = FALSE, warning = FALSE, echo = FALSE}
annot.all <- annot %>%
  select(transcript_id,
         sprot_Top_BLASTX_hit,
         sprot_Top_BLASTP_hit,
         gene_ontology_BLASTX,
         gene_ontology_BLASTP,
         Kegg,
         eggnog,
         Pfam) %>% 
  separate(transcript_id, into = c("pt1", "pt2", "gene", "isoform")) %>%
  unite(gene_id, pt1, pt2, gene, remove = FALSE) %>% 
  unite(transcript_id, pt1, pt2, gene, isoform) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot_geneName_BlastX", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
  separate(sprot2, sep = "=", into = c(NA, "sprot_geneFunction_BlastX")) %>% 
  separate(sprot_geneFunction_BlastX, sep = ";", into = c("sprot_geneFunction_BlastX", NA)) %>% 
  separate(sprot_Top_BLASTP_hit, into = c("sprot_geneName_BlastP", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
  separate(sprot2, sep = "=", into = c(NA, "sprot_geneFunction_BlastP")) %>% 
  separate(sprot_geneFunction_BlastP, sep = ";", into = c("sprot_geneFunction_BlastP", NA)) %>% 
  separate(gene_ontology_BLASTX, sep = "\\`", into = paste("GO_BlastX", 1:5, sep = "_"), extra = "drop", fill = "right") %>% 
  separate(gene_ontology_BLASTP, sep = "\\`", into = paste("GO_BlastP", 1:5, sep = "_"), extra = "drop", fill = "right") %>% 
  separate(Pfam, sep = "\\`", into = paste("Pfam", 1:5, sep = "_"), extra = "drop", fill = "right") %>% 
  as_tibble() %>% 
  pivot_longer(cols = 3:23, names_to = "dataSource", values_to = "annotation") %>% 
  drop_na(annotation)
```

## Differential expression analyses
### DE plotting scenario 1: 
### Single transcripts of interest from the DE results with boxplots
```{r, message = FALSE, warning = FALSE, echo = FALSE}
targetTrans <- "DN6338_c0_g1_i3"
lcpm %>%
  filter(transcript %in% targetTrans) %>% 
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "expression") %>% 
  separate(sample, into = c(NA, NA, "Tissue", "bird", NA)) %>% 
  mutate(bird = as.integer(bird)) %>% 
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "CONTROL", "LOW", "MODERATE", "HIGH")) %>% 
  ggplot(aes(y = expression, x = factor(SSgroup.virus.sac), fill = factor(Tissue))) +
  ylab("Log(Counts per million)") +
  xlab("Virus shedding group on day 1") +
  ggtitle(paste("Transcript: ", targetTrans)) +
  geom_boxplot() +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank())
```
  
### Single transcripts of interest from the DE results with scatterplots
The line is drawn as a Loess fuction and gray area shows 95% confidence intervals.
```{r, message = FALSE, warning = FALSE, echo = FALSE}
targetTrans <- "DN6338_c0_g1_i3"
lcpm %>%
  filter(transcript %in% targetTrans) %>% 
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "lcpm") %>% 
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>% 
  mutate(bird = as.integer(bird)) %>% 
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  ggplot(aes(x = log(virus.sac+0.01), y = lcpm, color = tissue)) +
  geom_point() +
  geom_smooth(method = "loess", span = 1.0) +
  ylab("Log(Counts per million)") +
  xlab("Log(Viral titer on day of sacrifice)") +
  ggtitle(paste("Transcript: ", targetTrans)) +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank())
```
  
### DE plotting scenario 2: 
### Heatmaps to reveal gene/transcript and sample clusters
```{r, warning = FALSE, message = FALSE, echo = FALSE, fig.height = 14, fig.width = 10, fig.align = "center"}
deGeneCounts.ileum <- lcpm.ileum %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript") %>%
  filter(transcript %in% dt.tib.ileum$transcript) %>% 
  column_to_rownames("transcript")

newNames <- colnames(deGeneCounts.ileum) %>% 
  as_tibble() %>% 
  separate(value, into = c(NA, NA, "tissue", "sample", NA)) %>% 
  mutate(group1 = group,
    group2 = ssgroup.virus.sac) %>% 
  unite("sample", 1:4, sep = "_")

colnames(deGeneCounts.ileum) <- newNames$sample
deGeneCounts.ileum <- deGeneCounts.ileum %>% 
  select(matches("(*C1_|*I1_)")) %>% 
  as.matrix()
  
## Set up palette
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
# Set up colour vector for celltype variable
col.diff <- c("black", "purple", "orange", "green")[covars.1$ssgroup.virus.sac]

# Plot the heatmap
heatmap.2(deGeneCounts.ileum,
          col = rev(morecols(50)),
          trace = "none", 
          main = "Group DE genes",
          margins = c(10, 10),
          ColSideColors = col.diff, scale ="row")
```
  
  
### DE plotting scenario 3: 
### Plotting transcripts/genes that cluster together
#### Step 1: Identify clusters
```{r, warning = FALSE, message = FALSE, fig.height = 6, fig.width = 8, echo = FALSE}
## 6 cluster scenario
hc <- hclust(dist(deGeneCounts.ileum))
focalCluster <- cutree(hc, k = 7) %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  rename(cluster = value) %>% 
  arrange(cluster) %>% 
  filter(cluster == "3")

plot(hc)
```

#### Step 2: Plot expression of a cluster
Gray lines are individual transcripts and the blue line is their mean value. This is for just one tissue type- the one that is used to build heatmaps and establish groupings (ileum in this case).  
```{r, echo = FALSE, warning = FALSE, message = FALSE}
targetTrans <- focalCluster$transcript

lcpm %>%
  filter(transcript %in% targetTrans) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "expression") %>%
  separate(sample, into = c(NA, NA, "Tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "CONTROL", "LOW", "MODERATE", "HIGH")) %>%
  arrange(SSgroup.virus.sac) %>%
  mutate(
    xVar = paste(SSgroup.virus.sac, "-", bird),
    xVar = factor(xVar, levels = unique(xVar))
  ) %>%
  select(transcript, Tissue, bird, expression, xVar) %>%
  filter(Tissue == "Ileum") %>%
  ggplot(aes(y = expression, x = xVar, group = transcript)) +
  ylab("Log(Counts per million)") +
  xlab("Bird") +
  geom_line(color = "gray") +
  stat_summary(
    aes(y = expression, group = 1),
    fun.y = mean,
    colour = "blue",
    geom = "line",
    group = 1, 
    lwd = 2
  ) +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
```
  
  
### Analysis of priori genes plotting scenario 1 
### Plotting all transcripts/genes with a given annotation: Categorical x axis, line plot

This example is for all genes with a SwissProt annotation for MX protein encoding genes. Note that these plots put all associated transcripts on one figure. They could also be broken apart into large multi-panel figures if preferable.
```{r, message = FALSE, warning = FALSE, echo = FALSE}
ctlInf.figure <- function(queryGene, ...) {
  
  ## Get list of transcripts associated with a gene name
  targetTrans <- annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, NA, NA), "\\^") %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  filter(SwissProt_GeneName == queryGene) %>% 
  distinct(transcript_id, .keep_all = TRUE)

  ## Create a vector of those transcripts
  targetTrans <- targetTrans$transcript_id
  
  ## Create the tibble that will be plotted 
  plot.dat <- lcpm %>%
  filter(transcript %in% targetTrans) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "expression") %>%
  separate(sample, into = c(NA, NA, "Tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  filter(group == "I1" | group == "C1") %>%
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "CONTROL", "LOW", "MODERATE", "HIGH")) %>%
  arrange(SSgroup.virus.sac) %>%
  mutate(
    xVar = paste(SSgroup.virus.sac, "-", bird),
    xVar = factor(xVar, levels = unique(xVar))
  ) %>%
  select(transcript, Tissue, bird, expression, xVar)
  
  ## Subset plotting tibble into bursa and ileum tibbles
  plot.dat.ileum <- plot.dat %>% filter(Tissue == "Ileum")
  plot.dat.bursa <- plot.dat %>% filter(Tissue == "Bursa")

  ## Generate plot
  plot <- ggplot(data = plot.dat.ileum, aes(y = expression, x = xVar, group = transcript)) +
  geom_line(color = "red", alpha = 0.1) +
  geom_line(data = plot.dat.bursa,
    aes(y = expression, x = xVar, group = transcript),
    color = "blue",
    alpha = 0.1
  ) +
  stat_summary(aes(y = expression, group = 1),
    fun.y = mean,
    colour = "red",
    geom = "line",
    group = 1,
    lwd = 2
  ) +
  stat_summary(data = plot.dat.bursa,
    aes(y = expression, group = 1),
    fun.y = mean,
    colour = "blue",
    geom = "line",
    group = 1,
    lwd = 2
  ) +
  ggtitle(paste("Transcripts annotated as", queryGene)) +
  geom_rect(aes(xmin = 0.4, xmax = 2.6, ymin = 8.5, ymax = 10.5
  ),
  color = "black", fill = "white") +
  annotate("Text", x = 1.5, y = 10, label = "Bursa", color = "blue", fontface = 2,size = 5
  ) +
  annotate("Text", x = 1.5, y = 9.0, label = "Ileum",
    color = "red", fontface = 2, size = 5
  ) +
  ylab("Log(Counts per million)") +
  xlab("Bird") +
  theme_classic(base_size = 14) +
  theme(legend.title = element_blank(),
        axis.text.x = element_text(angle = 45, hjust = 1))
  
  print(plot)
}

ctlInf.figure("MX")

```

### Analysis of priori genes plotting scenario 1 
### Plotting all transcripts/genes with a given annotation: Continuous x axis, scatterplot with Loess

```{r, message = FALSE, warning = FALSE, fig.height = 25, echo = FALSE}
targetTrans <- annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit) %>%
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, NA, NA), "\\^") %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  filter(SwissProt_GeneName == "MX") %>% 
  distinct(transcript_id, .keep_all = TRUE)

covars.sub <- covars %>% 
  filter(group == "I1" | group == "I3") %>% 
  select(bird, virus.sac) %>% 
  mutate(bird = sprintf("%02d", bird)) %>% 
  mutate(subSam = paste0("_", bird, "_"))
  
tmp.lcpm <- lcpm %>% 
  select(contains(covars.sub$subSam)) %>%
  mutate(transcript_id = lcpm$transcript) %>% 
  filter(transcript_id %in% targetTrans$transcript_id) %>% 
  pivot_longer(cols = starts_with("align"),
               names_to = "sample",
               values_to = "lcpm") %>% 
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>% 
  left_join(covars.sub)


for(gene in unique(tmp.lcpm$transcript_id)) {
  name <- paste0("plot.", gene)
  assign(name, tmp.lcpm %>%
      filter(transcript_id == gene) %>% 
      ggplot(aes(x = log(virus.sac), y = lcpm, color = tissue)) +
      geom_point() +
      geom_smooth(method = "loess", span = 1.5, se = FALSE) +
      ylab("Log(Counts per million)") +
      xlab("Log(Viral titer on day of sacrifice)") +
      ggtitle(gene) +
      theme_classic(base_size = 14) +
      theme(legend.title = element_blank(),
            axis.text.x = element_text(angle = 45, hjust = 1)))
}

grid.arrange(plot.DN110170_c0_g1_i1, plot.DN2085_c0_g2_i1,
             plot.DN2085_c0_g2_i2, plot.DN2085_c0_g2_i5,
             plot.DN2085_c0_g2_i6, plot.DN2085_c0_g2_i7,
             plot.DN2085_c0_g2_i10, plot.DN2085_c0_g2_i11,
             plot.DN2085_c0_g2_i12, plot.DN2085_c0_g2_i13,
             plot.DN2085_c0_g2_i14, plot.DN2085_c0_g2_i15,
             plot.DN2085_c0_g2_i16, nrow = 7)
```

