---
title: "LPAIV-infected and uninfected control groups - Ileum - Transcript"
author: "Jared J. Homola"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
This is an analysis of differential expression at the transcript level between treatment groups for control and infected blue-winged teal ileum samples.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(ggrepel)
library(RColorBrewer)
library(gridExtra)
library(kableExtra)
library(topGO)
library(clusterProfiler)
library(tidyverse)

## Load data
annot <- read.delim("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt <- read.table("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/rsem.isoform.counts.matrix", header = TRUE)
covars <- read.csv("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)
```


### Clean data
Bird 36 is missing a measurement for the weight covariate and bird 19 had improper tissue-based grouping, suggesting a potential sample mix up. They are both removed from the analysis. We also rename the transcripts, dropping the "Trinity_" prefix.
```{r, warning = FALSE, message = FALSE}
cnt <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Ileum"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript_id, pt1, pt2, gene, isoform)

```

### Set up model effects
```{r, warning = FALSE, message = FALSE}
bird <- as.factor(covars$bird)
sex <- as.factor(covars$sex)
age <- as.numeric(covars$age)
weight <- covars$wt_55
group <- recode(covars$group, C1 = "Ctl", C14 = "Ctl") %>% 
  as.factor()
pool <- as.factor(covars$Pool.Ileum)
```

### Prep data for analysis
Here we standardize expression levels across samples of varying depth by using the counts per million (CPM) and trimmed mean of M-values (TMM) methods. We also remove any transcripts that are expressed at few than 0.5 CPM in at least 25% of individuals.
```{r, warning = FALSE, message = FALSE}
#Convert to DGEList object
dge <- DGEList(counts=cnt)
dge$genes <- annot[match(rownames(dge$counts), annot$transcript_id),]

#CPM and log-CPM
cpm <- cpm(dge)
lcpm <- cpm(dge, log = TRUE)

#### Retain only transcripts expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge$counts==0) == length(cnt[1,]))
keep.exprs <- rowSums(cpm>0.5) >= length(cnt[1,])/4
sum(keep.exprs)

dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge$counts==0) == length(cnt[1,]))
dim(dge)

#TMM normalization
dge <- calcNormFactors(dge, method = "TMM")
```

### Plot MDS
Multidimensional scaling (MDS) plots are an ordination technique that lets us examine sample clustering based on overall gene expression levels.
```{r}
col.group = as.numeric(as.factor(paste0(group)))
plotMDS(lcpm, labels = paste0(bird, "_", group), col = col.group, cex = 0.75)
title(main="Treatment group")
```

# Differential expression analysis
### Establish design & contrast matrices
```{r}
## Establish design matrix
design = model.matrix(~ 0 +
                        group +
                        age +
                        sex + 
                        weight + 
                        pool)

colnames(design) <- gsub("group", "", colnames(design))

contr.matrix = makeContrasts(
  CtlvI1 = Ctl - I1,
  CtlvI3 = Ctl - I3,
  CtlvI5 = Ctl - I5,
  CtlvI14 = Ctl - I14,
  I1vI3 = I1 - I3,
  I3vI5 = I3 - I5,
  I5vI14 = I5 - I14,
  levels = colnames(design))
```

### Mean-variance trend and sample weight plots
```{r}
v <- voomWithQualityWeights(dge, design, plot = TRUE)
```

### Fitting the model
```{r}
vfit <- lmFit(v, design) 
vfit <- contrasts.fit(vfit, contrasts = contr.matrix)
tfit <- treat(vfit, lfc = 1.0)
plotSA(tfit)
```

### Count DE transcripts
For a transcript to be considered differentially expressed, we require a p-value of 0.1 with a false discovery rate correction and a log fold count difference of 1.
```{r}
dt <- decideTests(tfit, p.value = 0.1, adjust.method = "fdr")
summary(dt) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)

dt.tib <- as_tibble(dt, rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  mutate_at(vars(starts_with("C")), as.numeric) %>% 
  mutate_at(vars(starts_with("I")), as.numeric) %>% 
  filter_at(vars(CtlvI1:I5vI14), any_vars(. != 0))
```

### Histogram of P-values
```{r}
hist(tfit$p.value, main = "P-values comparing groups (eBayes)")
```


### Volcano plot
```{r, warning = FALSE, message = FALSE, fig.height = 10, fig.width = 7, fig.align = "center"}
tmp1 <- topTreat(tfit, coef = 1, n = Inf)
results1 <- mutate(tmp1, sig=ifelse(tmp1$adj.P.Val<0.1, "Sig", "Not Sig"))
p1 <- ggplot(results1, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp2 <- topTreat(tfit, coef = 2, n = Inf)
results2 <- mutate(tmp2, sig=ifelse(tmp2$adj.P.Val<0.1, "Sig", "Not Sig"))
p2 <- ggplot(results2, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp3 <- topTreat(tfit, coef = 3, n = Inf)
results3 <- mutate(tmp3, sig=ifelse(tmp3$adj.P.Val<0.1, "Sig", "Not Sig"))
p3 <- ggplot(results3, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 5") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp4 <- topTreat(tfit, coef = 4, n = Inf)
results4 <- mutate(tmp4, sig=ifelse(tmp4$adj.P.Val<0.1, "Sig", "Not Sig"))
p4 <- ggplot(results4, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 14") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp5 <- topTreat(tfit, coef = 5, n = Inf)
results5 <- mutate(tmp5, sig=ifelse(tmp5$adj.P.Val<0.1, "Sig", "Not Sig"))
p5 <- ggplot(results5, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 1 vs. Infected, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp6 <- topTreat(tfit, coef = 6, n = Inf)
results6 <- mutate(tmp6, sig=ifelse(tmp6$adj.P.Val<0.1, "Sig", "Not Sig"))
p6 <- ggplot(results6, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 3 vs. Infected, Day 5") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")


tmp7 <- topTreat(tfit, coef = 7, n = Inf)
results7 <- mutate(tmp7, sig=ifelse(tmp7$adj.P.Val<0.1, "Sig", "Not Sig"))
p7 <- ggplot(results7, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 5 vs. Infected, Day 14") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

grid.arrange(p1, p2, p3, p4, p5, p6, p7, nrow = 4)
```



### Heatmap
```{r, warning = FALSE, message = FALSE, fig.height = 18, fig.width = 9, fig.align = "center"}
# Get the gene names for DE genes
deGeneCounts <- lcpm %>%
  as_tibble(rownames = NA) %>%
  rownames_to_column("transcript") %>%
  filter(transcript %in% dt.tib$transcript) %>%
  column_to_rownames("transcript")

group2 <- group %>% 
  as_tibble() %>% 
  mutate(group = sub("I3", "I03", value)) %>% 
  mutate(group = sub("I1", "I01", group)) %>% 
  mutate(group = sub("I5", "I05", group)) %>% 
  mutate(group = sub("I014", "I14", group))

newNames1 <- colnames(deGeneCounts) %>%
  as_tibble() %>%
  separate(value, into = c(NA, NA, NA, "sample", NA)) %>%
  mutate(group = group2$group) %>%
  unite("sample", 2:1, sep = "_")

colnames(deGeneCounts) <- newNames1$sample
deGeneCounts <- deGeneCounts %>% dplyr::select(order(colnames(.))) %>%
  as.matrix()


heatmap.annot <- annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit) %>% 
  filter(transcript_id %in% dt.tib$transcript) %>% 
  separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
  separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>% 
  distinct(transcript_id, SwissProt_GeneName) %>% 
  filter(SwissProt_GeneName != ".")

deGeneCounts <- deGeneCounts %>% 
  as_tibble(rownames = NA) %>% 
  mutate(transcript = rownames(.)) %>% 
  rename(transcript_id = transcript) %>% 
  left_join(heatmap.annot) %>% 
  mutate(SwissProt_GeneName = replace_na(SwissProt_GeneName, ".")) %>% 
  unite(transcript, transcript_id:SwissProt_GeneName, sep = " - ") %>% 
  column_to_rownames("transcript") %>% 
  as.matrix()

## Set up palette
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)

hc <- hclust(as.dist(1-cor(t(deGeneCounts))))

# Plot the heatmap
heatmap.2(deGeneCounts,
          Colv = FALSE,
          Rowv = as.dendrogram(hc),
          col = rev(morecols(50)),
          trace = "none",
          colsep = c(9, 21, 33, 45),
          dendrogram = "row",
          density.info = "none",
          key = TRUE,
          #main = "Ileum - Gene level",
          margins = c(10, 15),
          scale ="row")
```

### Boxplots for differentially expressed transcripts
```{r message=FALSE, warning=FALSE}
lcpm.DE <- lcpm %>%
  as_tibble(rownames = NA) %>% 
  rownames_to_column("identifier") %>% 
  filter(identifier %in% dt.tib$transcript) %>%
  pivot_longer(cols = contains("_"),
               names_to = "sample",
               values_to = "lcpm") %>%
  separate(sample, into = c(NA, NA, "tissue", "bird", NA)) %>%
  mutate(bird = as.integer(bird)) %>%
  left_join(covars, by = "bird") %>%
  select(identifier, bird, lcpm, group) %>% 
  mutate(group = recode(group, C14 = "Control"),
         group = recode(group, C1 = "Control"),
         tmpID = ifelse(group == "Control", 'Control', 'Infected'))


aprioriPlotting <- function(target, ...) {
  annot.target <- annot %>%
    select(
      transcript_id,
      sprot_Top_BLASTX_hit) %>%
    filter(transcript_id == target) %>% 
    separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, NA, NA), "\\^")
  
  plot <- lcpm.DE %>%
    filter(identifier == target) %>%
    mutate(group = fct_relevel(group, "Control", "I1", "I3", "I5", "I14")) %>%
    ggplot(aes(x = group, y = lcpm)) +
    facet_grid(. ~ tmpID, scales = "free", space = "free") +
    ylab("Log2(Counts per million)") +
    xlab("Group") +
    geom_point(position = position_dodge(width=0.75), show.legend = FALSE) +
    geom_boxplot(alpha = 0.5) +
    geom_label_repel(aes(label = bird, fill = NULL), show.legend=FALSE) +
    theme_classic() +
    labs(title= paste0(target, " - ", annot.target[1,2])) +
    theme(legend.title = element_blank())
  
  print(plot)
}

for(trans in sort(unique(lcpm.DE$identifier))) {
  aprioriPlotting(trans)
}
```


### Annotations for differentially expressed transcripts
ns denotes non-significant transcripts for each comparison and numerical values are the log(fold count) difference
```{r message=FALSE, warning=FALSE}
tmp1a <- tmp1 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>%  mutate(comp = "CtlvI1")
tmp2a <- tmp2 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "CtlvI3")
tmp3a <- tmp3 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "CtlvI5")
tmp4a <- tmp4 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "CtlvI14")
tmp5a <- tmp5 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "I1vI3")
tmp6a <- tmp6 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "I3vI5")
tmp7a <- tmp7 %>% select(transcript_id, adj.P.Val, logFC, sprot_Top_BLASTX_hit, Kegg) %>% separate(sprot_Top_BLASTX_hit, into = c("SwissProt_GeneName", NA, NA, NA, NA, "sprot2", NA), "\\^") %>%   separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>%  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% mutate(comp = "I5vI14")

bind_rows(tmp1a, tmp2a, tmp3a, tmp4a, tmp5a, tmp6a, tmp7a) %>% 
  select(-SwissProt_GeneFunction, -KEGG_ID, -KO_ID) %>% 
  filter(transcript_id %in% dt.tib$transcript) %>% 
  mutate(logFoldCount = ifelse(adj.P.Val < 0.10, round(logFC, 2), "ns")) %>% 
  select(-adj.P.Val, -logFC) %>% 
  pivot_wider(names_from = comp, values_from = logFoldCount) %>% 
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

#### Transcript functions
```{r message=FALSE, warning=FALSE}
bind_rows(tmp1a, tmp2a, tmp3a, tmp4a, tmp5a, tmp6a, tmp7a) %>% 
  select(-comp, -adj.P.Val, -logFC) %>% 
  filter(transcript_id %in% dt.tib$transcript) %>% 
  distinct(.) %>% filter(SwissProt_GeneName != ".") %>% 
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```


### Gene Ontology Enrichment Analysis
```{r message=FALSE, warning=FALSE}
topGO.mappings <- readMappings("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/extData/GOdbTrans.txt", sep = "\t", IDsep = ",")

goEnrich <- function(targetComp, ...) {
  DE.results <- dt.tib %>% filter(!!sym(targetComp) != 0)
  
  all.genes <- sort(unique(as.character(rownames(tfit$p.value))))
  int.genes <- DE.results$transcript
  int.genes <- factor(as.integer(all.genes %in% int.genes))
  names(int.genes) = all.genes
  
  go.obj.BP <- new("topGOdata", ontology='BP',
              allGenes = int.genes,
              annot = annFUN.gene2GO,
              nodeSize = 5,
              gene2GO = topGO.mappings)

  go.obj.MF <- new("topGOdata", ontology='MF',
                 allGenes = int.genes,
                 annot = annFUN.gene2GO,
                 nodeSize = 5,
                 gene2GO = topGO.mappings)

  go.obj.CC <- new("topGOdata", ontology='CC',
                 allGenes = int.genes,
                 annot = annFUN.gene2GO,
                 nodeSize = 5,
                 gene2GO = topGO.mappings)
  
  results.BP <- runTest(go.obj.BP, algorithm = "elim", statistic = "fisher")

  results.tab.BP <- GenTable(object = go.obj.BP, 
                          elimFisher = results.BP, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "BP") %>% 
    mutate(Comparison = targetComp)
  
  results.MF <- runTest(go.obj.MF, algorithm = "elim", statistic = "fisher")
  
  results.tab.MF <- GenTable(object = go.obj.MF, 
                          elimFisher = results.MF, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "MF") %>% 
    mutate(Comparison = targetComp)
  
  results.CC <- runTest(go.obj.CC, algorithm = "elim", statistic = "fisher")
  
  results.tab.CC <- GenTable(object = go.obj.CC, 
                          elimFisher = results.CC, 
                          topNodes = 50) %>% 
    as_tibble() %>% 
    mutate(pVal = as.numeric(elimFisher)) %>% 
    filter(pVal < 0.01) %>% 
    mutate(Domain = "CC") %>% 
    mutate(Comparison = targetComp)
  
  rbind(results.tab.BP,
    results.tab.MF,
    results.tab.CC)
}

results <- list()

for(z in unique(names(dt.tib)[-1])) {
  if (dt.tib %>% 
      filter(!!sym(z) != 0) %>% 
      nrow(.) > 0)
  results[[length(results)+1]] <- goEnrich(z)
}

results.tib <- bind_rows(results, .id = "column_label") %>%
  select(-column_label) %>% 
  arrange(GO.ID, Comparison)

kable(results.tib) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```
