---
title: "Peak shedding level comparison - Ileum - Transcript"
author: "Jared J. Homola"
date: "4/17/2020"
output: html_document
---
This is an analysis of differential expression at the transcript level between supershedder groupings based on shedding rates measured on the day of sacrifice for infected blue-winged teal ileum samples.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(ggrepel)
library(RColorBrewer)
library(kableExtra)
library(gridExtra)
library(tidyverse)

setwd("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE manuscript/")

## Load data
annot <- read.delim("./extData/Trinotate.csv", header = TRUE, sep = "\t")
cnt <- read.table("./extData/rsem.isoform.counts.matrix", header = TRUE)
covars <- read.csv("./extData/BWTE54_SSgroup_Raw_Pool.csv", header = TRUE)
```


### Clean data
Bird 36 is missing a measurement for the weight covariate and bird 19 had improper tissue-based grouping, suggesting a potential sample mix up. They are both removed from the analysis. We also rename the transcripts, dropping the "Trinity_" prefix.
```{r, warning = FALSE, message = FALSE}
cnt <- cnt %>%
  select(
    -alignEstimateAbundance_BWTE_Ileum_36_S50,
    -alignEstimateAbundance_BWTE_Bursa_36_S31,-alignEstimateAbundance_BWTE_Ileum_19_S35,
    -alignEstimateAbundance_BWTE_Bursa_19_S14) %>%
  rownames_to_column("transcript") %>%
  separate(transcript, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript, pt1, pt2, gene, isoform) %>%
  column_to_rownames("transcript") %>% 
  select(contains("Ileum"))

covars <- covars %>%
  filter(!bird %in% c("36", "19")) %>%
  arrange(bird) %>%
  mutate(group = str_remove(group, "-"))

annot <- annot %>%
  separate(transcript_id, into = c(NA, "pt1", "pt2", "gene", "isoform")) %>%
  unite(transcript_id, pt1, pt2, gene, isoform)

```


### Set up model effects
```{r, warning = FALSE, message = FALSE}
bird <- as.factor(covars$bird)
sex <- as.factor(covars$sex)
age <- as.numeric(covars$age)
weight <- covars$wt_55
group <- recode(covars$group, C1 = "Ctl", C14 = "Ctl") %>% 
  as.factor()
pool <- as.factor(covars$Pool.Ileum)
ssgroup.virus.sac <- as.factor(covars$SSgroup.virus.sac)

covars.tib <- data.frame(bird,
                    sex,
                    age,
                    weight,
                    group,
                    pool,
                    ssgroup.virus.sac) %>% 
  as_tibble()
```

### Prep data for analysis
Here we standardize expression levels across samples of varying depth by using the counts per million (CPM) and trimmed mean of M-values (TMM) methods. We also remove any transcripts that are expressed at few than 0.5 CPM in at least 25% of individuals. 
```{r, warning = FALSE, message = FALSE}
#Convert to DGEList object
dge <- DGEList(counts=cnt)
dge$genes <- annot[match(rownames(dge$counts), annot$transcript_id),]

#CPM and log-CPM
cpm <- cpm(dge)
lcpm <- cpm(dge, log = TRUE)

#### Retain only transcripts expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge$counts==0) == length(cnt[1,]))
keep.exprs <- rowSums(cpm>0.5) >= length(cnt[1,])/4
sum(keep.exprs)

dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge$counts==0) == length(cnt[1,]))
dim(dge)

#TMM normalization
dge <- calcNormFactors(dge, method = "TMM")
```

### Subset data
Following normalization, we subset the data based on infection groups.
```{r, warning = FALSE, message = FALSE}
#Assign groups in DGElist
dge$samples$group <- covars.tib$group

### Divide out I1 and I3
dge.I1 <- dge[,grep('\\bI1\\b', dge$samples$group)]
dge.I3 <- dge[,grep('I3', dge$samples$group)]

covars.I1 <- covars.tib %>% filter(group == "I1")
covars.I3 <- covars.tib %>% filter(group == "I3")
```


# Differential expression analysis
For this analysis, we conduct two sets of differential expression analyses- one for each treatment group (I1 and I3).
```{r, warning = FALSE, message = FALSE,}
## Establish design matrix: I1
design.I1 = model.matrix(~ 0 +
                        factor(covars.I1$ssgroup.virus.sac) +
                        covars.I1$age +
                        covars.I1$sex + 
                        covars.I1$weight + 
                        covars.I1$pool)

colnames(design.I1) <- c("HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool3")


contr.matrix.I1 = makeContrasts(
  LvM = LOW - MODERATE,
  MvH = MODERATE - HIGH,
  LvH = LOW - HIGH,
  levels = colnames(design.I1))

## Mean-variance trend plots
v.I1 <- voomWithQualityWeights(dge.I1, design.I1, plot = FALSE)

#Fitting models
vfit.I1 <- lmFit(v.I1, design.I1) 
vfit.I1 <- contrasts.fit(vfit.I1, contrasts = contr.matrix.I1)
tfit.I1 <- treat(vfit.I1, lfc = 1.0)

## Identify DE genes
dt.I1 <- decideTests(tfit.I1, p.value = 0.1, adjust.method = "fdr")
dt.tib.I1 <- as_tibble(dt.I1, rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  rename(LvM.I1 = LvM,
         MvH.I1 = MvH,
         LvH.I1 = LvH)


## Establish design matrix: I3
design.I3 = model.matrix(~ 0 +
                        factor(covars.I3$ssgroup.virus.sac) +
                        covars.I3$age +
                        covars.I3$sex + 
                        covars.I3$weight + 
                        covars.I3$pool)

colnames(design.I3) <- c("HIGH", "LOW", "MODERATE",
                         "age", "sexM", "weight", "pool3")


contr.matrix.I3 = makeContrasts(
  LvM = LOW - MODERATE,
  MvH = MODERATE - HIGH,
  LvH = LOW - HIGH,
  levels = colnames(design.I3))

## Mean-variance trend plots
v.I3 <- voomWithQualityWeights(dge.I3, design.I3, plot = FALSE)

#Fitting models
vfit.I3 <- lmFit(v.I3, design.I3) 
vfit.I3 <- contrasts.fit(vfit.I3, contrasts = contr.matrix.I3)
tfit.I3 <- treat(vfit.I3, lfc = 1.0)

## Identify DE genes
dt.I3 <- decideTests(tfit.I3, p.value = 0.1, adjust.method = "fdr")
dt.tib.I3 <- as_tibble(dt.I3, rownames = NA) %>% 
  rownames_to_column("transcript") %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  rename(LvM.I3 = LvM,
         MvH.I3 = MvH,
         LvH.I3 = LvH)

## Bring it all together
dt.tib <- left_join(dt.tib.I1, dt.tib.I3) %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  filter_at(vars(LvM.I1:LvH.I3), any_vars(. != 0))
```

### Numbers of DE genes
```{r}
allResults <- rbind(
  as.data.frame(summary(dt.I1)),
  as.data.frame(summary(dt.I3))) %>% 
  filter(Var1 != "NotSig")

kable(allResults) %>%
  kable_styling("striped", full_width = F) %>%
  pack_rows("Infected 1", 1, 6) %>%
  pack_rows("Infected 3", 7, 12)

```

### Volcano plot
```{r, warning = FALSE, message = FALSE, fig.height = 10, fig.width = 10, fig.align = "center"}
tmp1 <- topTreat(tfit.I1, coef = 1, n = Inf)
results1 <- mutate(tmp1, sig=ifelse(tmp1$adj.P.Val<0.1, "Sig", "Not Sig"))
p1 <- ggplot(results1, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. Moderate shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")


tmp2 <- topTreat(tfit.I1, coef = 2, n = Inf)
results2 <- mutate(tmp2, sig=ifelse(tmp2$adj.P.Val<0.1, "Sig", "Not Sig"))
p2 <- ggplot(results2, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Moderate vs. High shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp3 <- topTreat(tfit.I1, coef = 3, n = Inf)
results3 <- mutate(tmp3, sig=ifelse(tmp3$adj.P.Val<0.1, "Sig", "Not Sig"))
p3 <- ggplot(results3, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. High shedding, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp4 <- topTreat(tfit.I3, coef = 1, n = Inf)
results4 <- mutate(tmp4, sig=ifelse(tmp4$adj.P.Val<0.1, "Sig", "Not Sig"))
p4 <- ggplot(results4, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. Moderate shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp5 <- topTreat(tfit.I3, coef = 2, n = Inf)
results5 <- mutate(tmp5, sig=ifelse(tmp5$adj.P.Val<0.1, "Sig", "Not Sig"))
p5 <- ggplot(results5, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Moderate vs. High shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp6 <- topTreat(tfit.I3, coef = 3, n = Inf)
results6 <- mutate(tmp6, sig=ifelse(tmp6$adj.P.Val<0.1, "Sig", "Not Sig"))
p6 <- ggplot(results6, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. High shedding, Day 3") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

grid.arrange(p1, p2, p3, p4, p5, p6, nrow = 2)
```


### Heatmap
```{r, warning = FALSE, message = FALSE, fig.height = 7, fig.width = 10, fig.align = "center"}
# Get the transcript names for DE transcripts
deTransCounts <- lcpm %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("transcript") %>%
  filter(transcript %in% dt.tib$transcript) %>% 
  column_to_rownames("transcript")

newNames <- colnames(deTransCounts) %>% 
  as_tibble() %>% 
  separate(value, into = c(NA, NA, "tissue", "sample", NA)) %>% 
  mutate(group = group) %>% 
  unite("sample", 1:3, sep = "_")

colnames(deTransCounts) <- newNames$sample
deTransCounts <- deTransCounts %>% 
  select(matches("(*I1|*I3)")) %>% 
  select(-contains("I14")) %>% 
  as.matrix()
  
## Set up palette
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)
# Set up colour vector for celltype variable
covars.I1I3 <- covars.tib %>% 
  filter(group == "I1" | group == "I3")
col.diff <- c("black", "purple", "orange", "green")[covars.I1I3$ssgroup.virus.sac]

# Plot the heatmap
heatmap.2(deTransCounts,
          col = rev(morecols(50)),
          trace = "none", 
          main = "Group DE transcripts",
          margins = c(5, 10),
          ColSideColors = col.diff, scale ="row")
```

### Annotations for differentially expressed transcripts
```{r message=FALSE, warning=FALSE}
annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit
  ) %>%
  rename(transcript = transcript_id) %>% 
  inner_join(dt.tib) %>% 
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
  separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  mutate_at(.vars = vars(LvM.I1:LvH.I3),
            .funs = ~ ifelse(. == 1, "up", .)) %>% 
  mutate_at(.vars = vars(LvM.I1:LvH.I3),
            .funs = ~ ifelse(. == -1, "down", .)) %>% 
  filter_at(vars(LvM.I1:LvH.I3), any_vars(. != 0)) %>% 
  distinct(transcript, .keep_all = TRUE) %>% 
  select(-SwissProt_GeneFunction) %>% 
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

#### Transcript functions
```{r message=FALSE, warning=FALSE}
annot %>%
  select(
    transcript_id,
    sprot_Top_BLASTX_hit,
    Kegg) %>%
  rename(transcript = transcript_id) %>% 
  inner_join(dt.tib) %>% 
  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
  separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>% 
  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% 
  distinct(transcript, SwissProt_GeneName, KEGG_ID, .keep_all = TRUE) %>% 
  select(transcript, SwissProt_GeneName, SwissProt_GeneFunction, KEGG_ID, KO_ID) %>% 
  filter(SwissProt_GeneName != ".") %>% 
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```
